### PR TITLE
Issues/181 - Add option to verify stub invocation with stub-runner

### PR DESCRIPTION
**Pull request by mchmielarz**

### CODE REVIEW

File: micro-deps/micro-deps-spring-test-config/src/main/groovy/com/ofg/infrastructure/stub/Interactions.groovy [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    void verifyThat(String collaboratorName, RequestPatternBuilder requestPatternBuilder, int count) {
+        mocks.get(collaboratorName).verifyThat(count, requestPatternBuilder)
+    }
+
+    void shutdown() {
+        mocks.each { it.value.shutdown() }
+    }
+
+    private WireMock createNewMock(String collaboratorName) {
+        if (isUnknown(collaboratorName)) {
```
Comment (nurkiewicz) [inline micro-deps/micro-deps-spring-test-config/src/main/groovy/com/ofg/infrastructure/stub/Interactions.groovy]: Maybe: `throwIfCollaboratorUnknown(collaboratorName)`?

**Comment by nurkiewicz** [comment]
This looks like a major new feature. Could you please document it somewhere on Wiki? E.g. here: https://github.com/4finance/micro-infra-spring/wiki/How-to-test-it

File: micro-deps/micro-deps-spring-test-config/src/main/groovy/com/ofg/infrastructure/stub/Interactions.groovy [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            throw new UnknownCollaboratorException(collaboratorName)
+        }
+        URL stubUrl = findCollaboratorStubUrl(collaboratorName)
+        return new WireMock(stubUrl.host, stubUrl.port)
+    }
+
+    private URL findCollaboratorStubUrl(String collaboratorName) {
+        String collaboratorPath = configurationResolver.dependencies[collaboratorName]['path']
+        def optionalStubUrl = stubRunning.findStubUrlByRelativePath(collaboratorPath)
+        if (optionalStubUrl.present) {
```
Comment (nurkiewicz) [inline micro-deps/micro-deps-spring-test-config/src/main/groovy/com/ofg/infrastructure/stub/Interactions.groovy]: `present` check is non-idiomatic. Would this work?

```groovy
optionalStubUrl.or{-> throw new MissingStubException(collaboratorName)}
```

File: micro-deps/micro-deps-spring-test-config/src/main/groovy/com/ofg/infrastructure/stub/Interactions.groovy [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    private boolean isUnknown(String collaboratorName) {
+        !configurationResolver.dependencies.containsKey(collaboratorName)
+    }
+
+}
+
+@CompileStatic
+class UnknownCollaboratorException extends RuntimeException {
```
Comment (nurkiewicz) [inline micro-deps/micro-deps-spring-test-config/src/main/groovy/com/ofg/infrastructure/stub/Interactions.groovy]: [`ServiceResolver`](https://github.com/4finance/micro-infra-spring/blob/master/micro-deps/micro-deps/src/main/groovy/com/ofg/infrastructure/discovery/ZookeeperServiceResolver.groovy#L46) simply uses `NoSuchElementException` in similar case.

File: stub-runner/stub-runner/src/main/groovy/com/ofg/stub/StubRunnerExecutor.groovy [hunk 4]
```diff
@@ -32,6 +33,20 @@ class StubRunnerExecutor {
...
+    /**
+     * Finds {@link URL} to a stubbed dependency by given {@code dependencyPath}.
+     *
+     * @param dependencyPath path taken from microservice configuration from dependency settings
+     *
+     * @return {@link URL} to stubbed dependency wrapped if found otherwise {@link Optional.absent()}
+     * @throws UnknownDependencyException in case when there is no stub defined for given {@code dependencyPath}
+     */
+    Optional<URL> getStubUrlByRelativePath(String dependencyPath) {
+        return stubServers.findResult(Optional.absent()) {
```
Comment (nurkiewicz) [inline stub-runner/stub-runner/src/main/groovy/com/ofg/stub/StubRunnerExecutor.groovy]: Alternative implementation that we find more readable:

```groovy
        return Optional.fromNullable(stubServers.find {
            return it.projectMetadata.projectRelativePath == dependencyPath
        }?.stubUrl)
```

File: micro-deps/micro-deps-spring-test-config/src/main/groovy/com/ofg/infrastructure/stub/Interactions.groovy [hunk 5]
```diff
@@ -0,0 +1,72 @@
+package com.ofg.infrastructure.stub
+
+import com.github.tomakehurst.wiremock.client.RequestPatternBuilder
+import com.github.tomakehurst.wiremock.client.WireMock
+import com.ofg.infrastructure.discovery.ServiceConfigurationResolver
+import com.ofg.stub.StubRunning
+import groovy.transform.CompileStatic
+
+class Interactions {
```
Comment (nurkiewicz) [inline micro-deps/micro-deps-spring-test-config/src/main/groovy/com/ofg/infrastructure/stub/Interactions.groovy]: I like this API, but what about making it even sexier? E.g.:

```groovy
Stub.of(collaboratorName).verityThat(requestPatternBuilder);
```

WDYT?

File: micro-deps/micro-deps-spring-test-config/src/test/groovy/com/ofg/infrastructure/stub/InteractionsSpec.groovy [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+        wireMock = new WireMock(LOCALHOST, mockServer.port())
+        wireMock.resetToDefaultMappings()
+    }
+
+    def 'should successfully verify interaction with stub'() {
+        given:
+            predefinedPongInteraction()
+            simulatedSinglePongInteraction()
+        when:
+            Interactions interactions = new Interactions(configurationResolver, stubRunning)
```
Comment (nurkiewicz) [inline micro-deps/micro-deps-spring-test-config/src/test/groovy/com/ofg/infrastructure/stub/InteractionsSpec.groovy]: Can we somehow add `Interactions` by default e.g. to instance of `com.ofg.infrastructure.base.MvcWiremockIntegrationSpec`?

File: micro-deps/micro-deps-spring-test-config/src/main/groovy/com/ofg/infrastructure/stub/Interactions.groovy (hunk 5 above)
Comment (mariuszs) [inline micro-deps/micro-deps-spring-test-config/src/main/groovy/com/ofg/infrastructure/stub/Interactions.groovy]: +1, `verityThat` should be static method for easier usage

Comment (mariuszs) [inline micro-deps/micro-deps-spring-test-config/src/main/groovy/com/ofg/infrastructure/stub/Interactions.groovy]: Second `verifyThat` still remember invocation from previous test.

**Pull request merged**